appcenter-data/ios/AppCenterReactNativeData/AppCenterReactNativeData.m

```diff
@@ -50,24 +61,102 @@ + (void)register {
-                 dict[kMSPartition] = document.partition;
-                 dict[kMSId] = document.documentId;
-                 dict[kMSETag] = document.eTag;
-                 dict[kMSLastUpdatedDate] = @([document.lastUpdatedDate timeIntervalSince1970] * 1000);
-                 dict[kMSIsFromDeviceCache] = [NSNumber numberWithBool:document.fromDeviceCache];
-                 return resolve(dict);
-             }];
+                     partition:partition
+                   readOptions:readOptions
+             completionHandler:[self dataCompletionHandler:@"Failed read" resolve:resolve reject:reject]];
+}
+
+RCT_EXPORT_METHOD(create:(NSString *)documentID
+                  partition:(NSString *)partition
+                  document: (NSDictionary *)documentMap
+                  writeOptions:(NSDictionary *)writeOptionsMap
+                  resolver:(RCTPromiseResolveBlock)resolve
+                  rejecter:(RCTPromiseRejectBlock)reject) {
+    MSWriteOptions *writeOptions = [self getWriteOptions:writeOptionsMap];
+    MSDictionaryDocument *document = [[MSDictionaryDocument alloc] initFromDictionary:documentMap];
+    [MSData createDocumentWithID:documentID
+                        document:document
+                       partition:partition
+                    writeOptions:writeOptions
+               completionHandler:[self dataCompletionHandler:@"Failed create" resolve:resolve reject:reject]];
+}
+
+RCT_EXPORT_METHOD(replace:(NSString *)documentID
+                  partition:(NSString *)partition
+                  document:(NSDictionary *)documentMap
+                  writeOptions:(NSDictionary *)writeOptionsMap
+                  resolver:(RCTPromiseResolveBlock)resolve
+                  rejecter:(RCTPromiseRejectBlock)reject) {
+    MSWriteOptions *writeOptions = [self getWriteOptions:writeOptionsMap];
+    MSDictionaryDocument *document = [[MSDictionaryDocument alloc] initFromDictionary:documentMap];
+    [MSData replaceDocumentWithID:documentID
+                         document:document
+                        partition:partition
+                     writeOptions:writeOptions
+                completionHandler:[self dataCompletionHandler:@"Failed replace" resolve:resolve reject:reject]];
+}
+
+RCT_EXPORT_METHOD(remove:(NSString *)documentID
+                  partition:(NSString *)partition
+                  writeOptions:(NSDictionary *)writeOptionsMap
+                  resolver:(RCTPromiseResolveBlock)resolve
+                  rejecter:(RCTPromiseRejectBlock)reject) {
+    MSWriteOptions *writeOptions = [self getWriteOptions:writeOptionsMap];
+    [MSData deleteDocumentWithID:documentID
+                       partition:partition
+                    writeOptions:writeOptions
+               completionHandler:[self dataCompletionHandler:@"Failed remove" resolve:resolve reject:reject]];
+}
+
+- (void (^)(MSDocumentWrapper * _Nonnull))dataCompletionHandler:(NSString *)errorCode
```

Comment:
An error code as `NSString` type doesn't look right. Error code is integer in native platforms, and you indeed have an error message string here.

Reply:
[`RCTPromiseRejectBlock`](https://github.com/facebook/react-native/blob/master/React/Base/RCTBridgeModule.h#L38) has a parameter `code`. I thought `errorCode` would make it more explicit. Should we rename it to `code` instead?

Reply:
What did we do for auth?

Reply:
We pass in a string directly in [AppCenterReactNativeAuth.m#L72](https://github.com/microsoft/appcenter-sdk-react-native/blob/eedd023cc1d4d5af190630581dfa21e0390e9195/appcenter-auth/ios/AppCenterReactNativeAuth/AppCenterReactNativeAuth.m#L72).

Reply:
@jackycodes 

Looks like is a confusion with the variable naming.

The terminology "error code" generally refers to code to identify a specific error case. In the context of Objective-C / Cocoa conventions, it's usually all-digit code like `12345`. That's why I think it may be confusing to use error message like "Failed remove" as error code. 

Rename `(NSString *)errorCode` to `(NSString *)errorMessage` would solve this problem.

Reply:
Please ignore my previous comment.

I found the type definition of `RCTPromiseRejectBlock` (https://github.com/facebook/react-native/blob/ff66600224e78fec5d0e902f8a035b78ed31a961/React/Base/RCTBridgeModule.h#L38). The first parameter is a string error code.

```
/**
 * Block that bridge modules use to reject the JS promise waiting for a result.
 * The error may be nil but it is preferable to pass an NSError object for more
 * precise error messages.
 */
typedef void (^RCTPromiseRejectBlock)(NSString *code, NSString *message, NSError *error);
```

An example usage of `RCTPromiseRejectBlock` can be found at https://github.com/facebook/react-native/blob/3724810d2168eb182db24acf9e741775df27ae13/Libraries/LinkingIOS/RCTLinkingManager.m#L100.

To follow the example, we can use keep this variable naming `errorCode` and pass in `"ReadFailed"`/`"CreateFailed"`/`"ReplaceFailed"`/`"RemoveFailed"`/etc as the first parameter in the caller.

Reply:
Sounds good, I’ll make a separate PR for Android